### PR TITLE
enhance gguf check

### PR DESCRIPTION
**Pull request by reidliu41**
**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [x] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Functional tests have been added, if nessessary.
- [ ] E2E Workflow tests have been added, if necessary.

### CODE REVIEW

**Comment by reidliu41** [comment]
Resolves #  if any invalid files under model dir, it will fail
```
$ ilab model list
Traceback (most recent call last):
  File "/Users/riddle/ai/projects/instructlab-dev/venv/bin/ilab", line 8, in <module>
    sys.exit(ilab())
             ^^^^^^
  File "/Users/riddle/ai/projects/instructlab-dev/venv/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/riddle/ai/projects/instructlab-dev/venv/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/Users/riddle/ai/projects/instructlab-dev-2024/venv/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/riddle/ai/projects/instructlab-dev-2024/venv/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/riddle/ai/projects/instructlab-dev/venv/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/riddle/ai/projects/instructlab-dev/venv/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/riddle/ai/projects/instructlab-dev/instructlab/src/instructlab/clickext.py", line 306, in wrapper
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/Users/riddle/ai/projects/instructlab-dev/instructlab/src/instructlab/model/list.py", line 41, in model_list
    if entry.is_file() and is_model_gguf(entry):
                           ^^^^^^^^^^^^^^^^^^^^
  File "/Users/riddle/ai/projects/instructlab-dev/instructlab/src/instructlab/model/backends/backends.py", line 151, in is_model_gguf
    first_four_bytes_int = int(struct.unpack("<I", first_four_bytes)[0])
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
struct.error: unpack requires a buffer of 4 bytes
```

File: src/instructlab/model/backends/backends.py [hunk 1]
```diff
@@ -95,12 +95,19 @@ def is_model_gguf(model_path: pathlib.Path) -> bool:
...
+            # Convert the first four bytes to an integer
+            first_four_bytes_int = int(struct.unpack("<I", first_four_bytes)[0])
+        except struct.error as e:
+            logger.error(f"Failed to unpack the first four bytes of {model_path}: {e}")
+            return False
 
         return first_four_bytes_int == GGUF_MAGIC
     except IsADirectoryError as exc:
-        logger.debug(f"GGUF Path is a directory, returning {exc}")
+        logger.debug(f"GGUF Path  {model_path} is a directory, returning {exc}")
```
Comment (danmcp) [inline src/instructlab/model/backends/backends.py:107]: nit: There is an extra space after Path

File: src/instructlab/model/backends/backends.py [hunk 2]
```diff
@@ -95,12 +95,19 @@ def is_model_gguf(model_path: pathlib.Path) -> bool:
...
+        return False
+    except Exception as exc:
+        logger.debug(f"Error reading file {model_path}: {exc}")
```
Comment (danmcp) [inline src/instructlab/model/backends/backends.py:110]: What's the logic for this being debug and the struct.error above is error?

File: src/instructlab/model/backends/backends.py [hunk 3]
```diff
@@ -95,12 +95,19 @@ def is_model_gguf(model_path: pathlib.Path) -> bool:
         with model_path.open("rb") as f:
             first_four_bytes = f.read(4)
 
-        # Convert the first four bytes to an integer
-        first_four_bytes_int = int(struct.unpack("<I", first_four_bytes)[0])
+        try:
+            # Convert the first four bytes to an integer
+            first_four_bytes_int = int(struct.unpack("<I", first_four_bytes)[0])
+        except struct.error as e:
+            logger.error(f"Failed to unpack the first four bytes of {model_path}: {e}")
```
Comment (danmcp) [inline src/instructlab/model/backends/backends.py:102]: This message might be a little obscure for the error level?  Maybe explain what we are looking for?

**Review by danmcp — CHANGES_REQUESTED** [review]

File: src/instructlab/model/backends/backends.py [hunk 4]
```diff
@@ -95,12 +95,20 @@ def is_model_gguf(model_path: pathlib.Path) -> bool:
         with model_path.open("rb") as f:
             first_four_bytes = f.read(4)
 
-        # Convert the first four bytes to an integer
-        first_four_bytes_int = int(struct.unpack("<I", first_four_bytes)[0])
+        try:
+            # Convert the first four bytes to an integer
+            first_four_bytes_int = int(struct.unpack("<I", first_four_bytes)[0])
+        except struct.error as e:
```
Comment (leseb) [inline src/instructlab/model/backends/backends.py:101]:
```suggestion
        except struct.error as exc:
```
Just to stay consistent with other `except` blocks.

**Review by leseb — CHANGES_REQUESTED** [review]

File: src/instructlab/model/backends/backends.py [hunk 5]
```diff
@@ -95,12 +95,20 @@ def is_model_gguf(model_path: pathlib.Path) -> bool:
...
             first_four_bytes = f.read(4)
 
-        # Convert the first four bytes to an integer
-        first_four_bytes_int = int(struct.unpack("<I", first_four_bytes)[0])
+        try:
+            # Convert the first four bytes to an integer
+            first_four_bytes_int = int(struct.unpack("<I", first_four_bytes)[0])
+        except struct.error as exc:
+            logger.debug(f"Failed to unpack the first four bytes of {model_path}. "
+                         "The file might not be a valid GGUF file or is corrupted: {e}")
```
Comment (leseb) [inline src/instructlab/model/backends/backends.py:103]:
```suggestion
                         "The file might not be a valid GGUF file or is corrupted: {exc}")
```

File: src/instructlab/model/backends/backends.py [hunk 6]
```diff
@@ -95,12 +95,20 @@ def is_model_gguf(model_path: pathlib.Path) -> bool:
...
             first_four_bytes = f.read(4)
 
-        # Convert the first four bytes to an integer
-        first_four_bytes_int = int(struct.unpack("<I", first_four_bytes)[0])
+        try:
+            # Convert the first four bytes to an integer
+            first_four_bytes_int = int(struct.unpack("<I", first_four_bytes)[0])
+        except struct.error as exc:
+            logger.debug(f"Failed to unpack the first four bytes of {model_path}. "
+                         "The file might not be a valid GGUF file or is corrupted: {exe}")
```
Comment (leseb) [inline src/instructlab/model/backends/backends.py:103]:
```suggestion
                         "The file might not be a valid GGUF file or is corrupted: {exc}")
```

**Review by leseb — COMMENTED** [review]
Time to squash your commits, thanks!

File: README.md [hunk 7]
```diff
@@ -158,6 +158,8 @@ For an overview of the full workflow, see the [workflow diagram](./docs/workflow
 
       > **NOTE**: Make sure your system Python build is `Mach-O 64-bit executable arm64` by using `file -b $(command -v python)`,
       > or if your system is setup with [pyenv](https://github.com/pyenv/pyenv) by using the `file -b $(pyenv which python)` command.
+      >
```
Comment (leseb) [inline README.md:161]: Wrong rebase?

**Review by leseb — COMMENTED** [review]
Nit: there is a typo in your commit message 'upate typo'.

File: README.md (hunk 7 above)
Comment (reidliu41) [reply README.md:161]: seems that, any methods to fix it?

Comment (leseb) [reply README.md:161]: Just rebase against main.

**Comment by leseb** [comment]
@Mergifyio rebase

File: README.md (hunk 7 above)
Comment (leseb) [reply README.md]: I just did it using the bot but you still need to update your commit message and push again, thanks!

File: src/instructlab/model/backends/backends.py [hunk 8]
```diff
@@ -95,12 +95,24 @@ def is_model_gguf(model_path: pathlib.Path) -> bool:
...
+                f"The file might not be a valid GGUF file or is corrupted: {exc}"
+            )
+            return False
 
         return first_four_bytes_int == GGUF_MAGIC
     except IsADirectoryError as exc:
-        logger.debug(f"GGUF Path is a directory, returning {exc}")
+        logger.debug(f"GGUF Path {model_path} is a directory, returning {exc}")
+        return False
+    except OSError as exc:
```
Comment (leseb) [inline src/instructlab/model/backends/backends.py:112]: Sorry I just realized that the inner try/except is not needed, given that we are already in a try/except block you can just append this in here:

```
        except struct.error as exc:
            logger.debug(
                f"Failed to unpack the first four bytes of {model_path}. "
                f"The file might not be a valid GGUF file or is corrupted: {exc}"
            )
```

Thanks!

**Review by leseb — COMMENTED** [review]
Can you share an output now with the latest version of the code? Thanks

**Comment by reidliu41** [comment]
~~~
$ touch /Users/xx/.cache/instructlab/models/testfile  # create a simple file

$ ilab model list
+------------------------------+---------------------+--------+
| Model Name                   | Last Modified       | Size   |
+------------------------------+---------------------+--------+
| merlinite-7b-lab-Q4_K_M.gguf | 2024-09-14 06:45:33 | 4.1 GB |
| test_namespace/testlab_model | 2024-09-20 11:44:07 | 48.0 B |
+------------------------------+---------------------+--------+
~~~

**Comment by leseb** [comment]
> ```
> $ touch /Users/xx/.cache/instructlab/models/testfile  # create a simple file
> 
> $ ilab model list
> +------------------------------+---------------------+--------+
> | Model Name                   | Last Modified       | Size   |
> +------------------------------+---------------------+--------+
> | merlinite-7b-lab-Q4_K_M.gguf | 2024-09-14 06:45:33 | 4.1 GB |
> | test_namespace/testlab_model | 2024-09-20 11:44:07 | 48.0 B |
> +------------------------------+---------------------+--------+
> ```

Thanks, same with debug please? `ilab -v model list`.

**Comment by reidliu41** [comment]
> > ```
> > $ touch /Users/xx/.cache/instructlab/models/testfile  # create a simple file
> > 
> > $ ilab model list
> > +------------------------------+---------------------+--------+
> > | Model Name                   | Last Modified       | Size   |
> > +------------------------------+---------------------+--------+
> > | merlinite-7b-lab-Q4_K_M.gguf | 2024-09-14 06:45:33 | 4.1 GB |
> > | test_namespace/testlab_model | 2024-09-20 11:44:07 | 48.0 B |
> > +------------------------------+---------------------+--------+
> > ```
> 
> Thanks, same with debug please? `ilab -v model list`.

```
$ ilab -v  model list
Parameters:

DEBUG 2024-09-24 21:01:31,641 instructlab.model.backends.backends:157: Failed to unpack the first four bytes of /Users/xx/.cache/instructlab/models/testfile. The file might not be a valid GGUF file or is corrupted: unpack requires a buffer of 4 bytes
```

**Review by leseb — COMMENTED** [review]
Last thing! Your commit message still has a typo in it, "upate typo", thanks!

**Review by leseb — APPROVED** [review]
Thanks!

**Review by danmcp — APPROVED** [review]

**Pull request merged**